server/sonar-db-dao/src/main/java/org/sonar/db/purge/PurgeDao.java

```diff
@@ -171,6 +180,55 @@ private static void deleteProject(String rootUuid, PurgeMapper mapper, PurgeComm
     commands.deleteWebhookDeliveries(rootUuid);
   }
 
+  /**
+   * Delete the non root components (ie. neither project nor view) from the specified collection of {@link ComponentDto}
+   * and data from their child tables.
+   * <p>
+   *   This method has no effect when passed an empty collection or only root components.
+   * </p>
+   */
+  public PurgeDao deleteNonRootComponents(DbSession dbSession, Collection<ComponentDto> components) {
+    Set<ComponentDto> nonRootComponents = components.stream().filter(PurgeDao::isNotRoot).collect(MoreCollectors.toSet());
+    if (nonRootComponents.isEmpty()) {
+      return this;
+    }
+
+    PurgeProfiler profiler = new PurgeProfiler();
+    PurgeCommands purgeCommands = new PurgeCommands(dbSession, profiler);
+    deleteNonRootComponents(nonRootComponents, purgeCommands);
+
+    return this;
+  }
+
+  private static void deleteNonRootComponents(Set<ComponentDto> nonRootComponents, PurgeCommands purgeCommands) {
+    List<IdUuidPair> modulesOrSubviews = nonRootComponents.stream()
+      .filter(PurgeDao::isModuleOrSubview)
+      .map(PurgeDao::toIdUuidPair)
+      .collect(MoreCollectors.toList());
+    purgeCommands.deleteByRootAndModulesOrSubviews(modulesOrSubviews);
+    List<String> nonRootComponentUuids = nonRootComponents.stream().map(ComponentDto::uuid).collect(MoreCollectors.toList(nonRootComponents.size()));
+    purgeCommands.deleteComponentMeasures(nonRootComponentUuids);
+    purgeCommands.deleteFileSources(nonRootComponents.stream().filter(PurgeDao::isFile).map(ComponentDto::uuid).collect(MoreCollectors.toList()));
+    purgeCommands.deleteIssues(nonRootComponentUuids);
+    purgeCommands.deleteComponents(nonRootComponentUuids);
+  }
+
+  private static boolean isNotRoot(ComponentDto dto) {
+    return !(SCOPE_PROJECT.equals(dto.scope()) && QUALIFIERS_PROJECT_VIEW.contains(dto.qualifier()));
```

Comment:
indeed the algorithm is different, it either relies on `module_uuid` in addition to scope (for method `isRootProject`) or solely on `uuid_path` (for method `isRoot`).

I'd rather keep to the column I already use in the methodfor other parts of the algorithm (eg. to identify modules/subview or files). It makes the whole thing simpler because more consistent.